### ISSUE TITLE
Dev Containers based setup of prerequisites

### ISSUE THREAD

**Issue by bitshop** [aws-ia/terraform-aws-eks-blueprints#1605]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

In order to explore the repo and terraform easier for those that don't currently use terraform I'd like to see a Dev Container file, this allows GitHub CodeSpaces or Visual Studio Code to set everything up with a few clicks. 

Build Dev Container folder/files to provision prerequisites. 

Note: Since I'm working on testing V5 my upcoming PR is against what appears to be that branch instead of main, I hope that's acceptable, if not I can make another PR to main as it's only a few files. 

There are alternatively platforms, however this repo is hosted in github so their standard seems to make the most sense. The alternatively development container standards can co-exist so there is no harm in including this.

**Comment by bitshop** [comment]
@bryantbiggs What's the logic to not adding this? Is there another container solution you prefer?

**Comment by bryantbiggs** [comment]
I don't think the added overhead of adding support for this, and any other solutions similar to it, are worthwhile for the project. To run the examples there are at most 3 CLIs that are required - Terraform, awscli, and kubectl. If you wish to install and run these from within a container, that is ok, but I don't think we need to add support for devcontainers specifically to this project

**Comment by bitshop** [comment]
Isn't the goal of the repo to be able to quickly learn, test and experiment? 

The PR allows me to in github simply click CodeSpaces and have terraform running - When people who don't use Terraform see the repo and try to decide between CDK and Terraform this is a quick solution to experiment - An exact use case a client of mine had and hence the setup for them. Enough corporate laptops are locked down that installs add a challenge, potentially an approval of new software (even if popular), etc. vs a click on GitHub. For files that should almost never change that seems very reasonable to provide (on most repos, not just this one).  

If you want me to commit to maintaining this subdirectory longer term that's fine, I'm confident it's rarely going to change.

**Comment by bryantbiggs** [comment]
The goal of the Blueprints projects are multi-faceted. I think its safe for us, the maintainers, to assume that users will have the 3 CLIs listed installed locally. 

In terms of the scenario you provided - if the laptops are locked down and restricted, how are AWS credentials typically handled in the sense of a devcontainer (I am not a user of devcontainers nor I have used them in the past)? And likewise, how would you handle the Terraform statefile in a devcontainer?

**Comment by bitshop** [comment]
I made a quick rough video of how easy it is to explore / test the system for someone new: https://youtu.be/3GEvLz4YNlU

AWS Creds: Environment variables is one method, I demonstrated that quickly in the video - I'll write more at end.. 

State: Like any terraform deployment by default it would be stored local, not in the container, however local varies:
- Laptop: On your disk, outside of the container
- EC2 instance w/ containers: On the host volume where you clone repo to
- CodeSpaces: In your code space, and persisted after the container is stopped or rebuilt. Deleting CodeSpace would delete the state files. 

Like any system the backends would be better to store the state files in, at least that's my understanding. 

-- May be of no interest: Elaborating on AWS Creds: Often for developers we give the EC2 instance a profile which has similar permissions to deployed DEV tier and then each dev can launch a dev container on the host and edit in VS Code locally - we restrict who has access to the host, we do lose the user name in CloudTrail but this makes developers lives pretty easy and dev containers, like any container, removes the it works here topics.
